### PR TITLE
test(drawer): playwright refactor

### PR DESCRIPTION
**Pull request by stephenogorman**
### Proposed behaviour

Refactor Drawer tests from Cypress to Playwright

### Current behaviour

Tests are currently Cypress

### Checklist

- [x] Commits follow our style guide
- [ ] Related issues linked in commit messages if required
- [ ] Screenshots are included in the PR if useful
- [ ] All themes are supported if required
- [ ] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [x] Playwright automation tests added or updated if required
- [ ] Storybook added or updated if required
- [ ] Translations added or updated (including creating or amending translation keys table in storybook) if required
- [ ] Typescript `d.ts` file added or updated if required
- [ ] Related docs have been updated if required

#### QA

- [ ] Tested in CodeSandbox/storybook
- [ ] Add new Cypress test coverage if required
- [ ] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Additional context

N/A

### Testing instructions
- [ ] Run Drawer tests in Playwright Test Runner to check if the `drawer.pw.tsx` file passed
- [ ] Run all other tests in Playwright Test Runner to check none of the other `*.pw.tsx` files have regressed

### CODE REVIEW

File: src/components/drawer/drawer.pw.tsx [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+  });
+
+  test("should render component with custom expandedWidth", async ({
+    mount,
+    page,
+  }) => {
+    await mount(<DrawerCustom expandedWidth="65%" />);
+
+    await assertCssValueIsApproximately(
+      drawer(page).locator("div").nth(0),
```
Comment (robinzigmond) [inline src/components/drawer/drawer.pw.tsx:178]:
```suggestion
      drawer(page).locator("div").first(),
```

**suggestion** (non-blocking): `first` exists in Playwright as well as in Cypress. Doesn't really matter much but `first()` is a bit more immediately obvious what's happening.

File: src/components/drawer/drawer.pw.tsx [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+  }) => {
+    await mount(<DrawerCustom backgroundColor="#FF0000" />);
+
+    const color = "rgb(255, 0, 0)";
+
+    await expect(drawer(page).locator("div").nth(0)).toHaveCSS(
+      "background-color",
+      color
+    );
+    await expect(drawer(page).locator("div").nth(0)).toHaveCSS(
```
Comment (robinzigmond) [inline src/components/drawer/drawer.pw.tsx:196]: **suggestion** (non-blocking): combining a couple of comments made on earlier tests:
1) the `.nth(0)` could be `.first()`
2) you could extract `draw(page).locator("div").first()` (or however you choose to write it) to a separate variable, to make the tests less repetitive and more descriptive.

File: src/components/drawer/drawer.pw.tsx [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+    mount,
+    page,
+  }) => {
+    await mount(<DrawerCustom height="75%" />);
+
+    await expect(drawer(page)).toHaveAttribute("height", "75%");
+  });
+
+  test("should render component with custom title", async ({ mount, page }) => {
+    await mount(<DrawerCustom title="cypress_title" />);
```
Comment (robinzigmond) [inline src/components/drawer/drawer.pw.tsx:212]: **nitpick** (non-blocking): again I think it's better not to use the word "cypress" in the custom title now

File: src/components/drawer/drawer.pw.tsx [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+          height: "400px",
+        }}
+      >
+        <DrawerCustomFooterHeader />
+      </div>
+    );
+
+    const footer = stickyFooter(page);
+    await footer.scrollIntoViewIfNeeded();
+    expect(await stickyFooter(page).count()).toBeGreaterThan(0);
```
Comment (robinzigmond) [inline src/components/drawer/drawer.pw.tsx:253]: **comment** (non-blocking): I don't think this `count` test is needed - it won't pass `toBeVisible()` below if it's not actually found in the DOM!

File: src/components/drawer/drawer.pw.tsx [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+  test("should render component without toggle control when showControls prop is falsy", async ({
+    mount,
+    page,
+  }) => {
+    await mount(<DrawerCustom />);
+
+    expect(await drawerToggle(page).count()).toEqual(0);
+  });
+
+  test("should render component with custom footer", async ({
```
Comment (robinzigmond) [inline src/components/drawer/drawer.pw.tsx:237]: **comment** (non-blocking): I admit I'm a bit confused by this test, and the changes that have happened to it from the Cypress version to your Playwright one.

The original Cypress one just asserted 2 things - that the sticky footer existed, but was not visible. As you've written it in Playwright, you're now scrolling it into view and checking that it's then visible, which the original test wasn't doing.

I think I understand what's happening, but I'd prefer to assert that the footer is *not* visible before the `scrollIntoViewIfNeeded` call, which would then leaving it (as far as I can see) testing basically the same thing(s) as before?

File: src/components/drawer/drawer.pw.tsx [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+      </div>
+    );
+
+    const footer = stickyFooter(page);
+    await footer.scrollIntoViewIfNeeded();
+    await expect(
+      drawer(page).locator("div").locator("div").locator("div").nth(0)
+    ).toHaveText("Drawer title");
+    await expect(
+      drawer(page).locator("div").locator("div").locator("div").nth(0)
```
Comment (robinzigmond) [inline src/components/drawer/drawer.pw.tsx:294]: **suggestion** (non-blocking): again, the result of this complex locator call could be extracted to a variable with a descriptive name

File: src/components/drawer/drawer.pw.tsx [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+    ["expanded", 0],
+    ["not expanded", 1],
+  ] as [string, number][]).forEach(([state, times]) => {
+    test(`should pass accessibility tests when chevron is ${state}`, async ({
+      mount,
+      page,
+    }) => {
+      await mount(<DrawerCustom showControls />);
+
+      for (let i = 0; i < times; i++) {
```
Comment (robinzigmond) [inline src/components/drawer/drawer.pw.tsx:351]: **suggestion** (non-blocking): as mentioned in one of the other tests, this loop can really just be an `if`, which would also avoid the lint warning.

File: src/components/drawer/drawer.pw.tsx [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+        await drawerToggle(page).click();
+        await expect(drawerSidebar(page)).toBeVisible();
```
Comment (robinzigmond) [inline src/components/drawer/drawer.pw.tsx:353]: **question** (non-blocking): do we need this visibility assertion? It's not there in the equivalent Cypress tests, and doesn't relate to the accessibility test.

Or is it something we need in Playwright to make sure it waits long enough for the sidebar to become visible? (I have no idea what the answer to that is.)

File: src/components/drawer/drawer.pw.tsx [hunk 9]
```diff
@@ -0,0 +1,488 @@
...
+    test(`should pass accessibility tests when animation time is set to ${animationDuration}`, async ({
+      mount,
+      page,
+    }) => {
+      await mount(
+        <DrawerCustom showControls animationDuration={animationDuration} />
+      );
+
+      await drawerToggle(page).click();
+      await expect(drawerSidebar(page)).toBeVisible();
```
Comment (robinzigmond) [inline src/components/drawer/drawer.pw.tsx:370]: **question** (non-blocking): same question here about whether the visibility assertion is actually needed or relevant to this particular test.

**Review by robinzigmond — DISMISSED** [review]

File: src/components/drawer/components.test-pw.tsx [hunk 10]
```diff
@@ -0,0 +1,173 @@
+import React from "react";
+
+import Drawer, { DrawerProps } from ".";
+import Typography from "../typography";
+import { Checkbox } from "../checkbox";
+import Box from "../box";
+import Button from "../button";
+
+export const DrawerCustom = (props: Partial<DrawerProps>) => {
```
Comment (ZhuoyuJin) [inline src/components/drawer/components.test-pw.tsx:9]: could you delete "DrawerCustom" from drawer-test.stories.tsx file please

File: cypress/locators/drawer/index.js [hunk 11]
```diff
@@ -1,8 +0,0 @@
-import { DRAWER, DRAWER_SIDEBAR, DRAWER_TOGGLE } from "./locators";
-
-// component preview locators
-export const drawerToggle = () => cy.get(DRAWER_TOGGLE);
-export const drawer = () => cy.get(DRAWER);
-export const drawerSidebar = () => cy.get(DRAWER_SIDEBAR);
-export const drawerSidebarContentInnerElement = (index) =>
-  drawerSidebar().find("li").eq(index);
```
Comment (ZhuoyuJin) [inline cypress/locators/drawer/index.js:8]: Is it necessary that we delete the index.js file?

**Review by ZhuoyuJin — CHANGES_REQUESTED** [review]

**Review by ZhuoyuJin — APPROVED** [review]

**Review by robinzigmond — APPROVED** [review]

**Review by ZhuoyuJin — APPROVED** [review]

**Review by robinzigmond — APPROVED** [review]

**Pull request merged**